### PR TITLE
aarch64 build support

### PR DESCRIPTION
**Pull request by peak3d**
aarch64 build support
## Description
- upgrade autotools for aarch64 depends build
- add aarch64 to the cmake build process
  ## Motivation and Context
  support aarch64 depends build and kodi cmake build
## How Has This Been Tested?

cross compiled on ubuntu x86-64 14.04 lts
## Types of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Review by wsnipex — APPROVED** [review]
imho good to go

**Comment by wsnipex** [comment]
jenkins build this please

File: project/cmake/scripts/linux/Macros.cmake [hunk 1]
```diff
@@ -32,18 +32,19 @@ function(core_link_library lib wraplib)
   endif()
 
   get_filename_component(dir ${wraplib} PATH)
-  add_custom_command(OUTPUT ${CMAKE_BINARY_DIR}/${wraplib}-${ARCH}${CMAKE_SHARED_MODULE_SUFFIX}
```
Comment (fetzerch) [inline project/cmake/scripts/linux/Macros.cmake]: this reverts https://github.com/xbmc/xbmc/pull/10614

File: project/cmake/scripts/linux/Macros.cmake [hunk 2]
```diff
@@ -32,18 +32,19 @@ function(core_link_library lib wraplib)
...
                      ARGS    -Wl,--whole-archive
                              "${link_lib}" ${extra_libs}
                              -Wl,--no-whole-archive -lm
+                             -Wl,-soname,${wraplib}-${ARCH}${CMAKE_SHARED_MODULE_SUFFIX}
                              -shared -o ${CMAKE_BINARY_DIR}/${wraplib}-${ARCH}${CMAKE_SHARED_MODULE_SUFFIX}
                              ${export}
                      DEPENDS ${target} wrapper.def wrapper)
 
   get_filename_component(libname ${wraplib} NAME_WE)
-  add_custom_target(wrap_${libname} ALL DEPENDS ${CMAKE_BINARY_DIR}/${wraplib}-${ARCH}${CMAKE_SHARED_MODULE_SUFFIX})
```
Comment (fetzerch) [inline project/cmake/scripts/linux/Macros.cmake]: same

File: tools/buildsteps/android-arm64-v8a/configure-xbmc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+WORKSPACE=${WORKSPACE:-$( cd $(dirname $0)/../../.. ; pwd -P )}
+XBMC_PLATFORM_DIR=android
+. $WORKSPACE/tools/buildsteps/defaultenv
+
+if [ "$CMAKE_BUILD" = true ]; then
```
Comment (fetzerch) [inline tools/buildsteps/android-arm64-v8a/configure-xbmc]: you can drop the non cmake branch, see https://github.com/xbmc/xbmc/pull/10649

File: tools/buildsteps/android-arm64-v8a/make-binary-addons [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ALL_BINARY_ADDONS_BUILT="1"
+#only build binary addons when requested by env/jenkins
+if [ "$BUILD_BINARY_ADDONS" == "true" ]
+then
+  for addon in $BINARY_ADDONS
+  do
+    echo "building $addon"
+    git clean -xffd $WORKSPACE/$BINARY_ADDONS_ROOT/$addon
+
+    if [ "$CMAKE_BUILD" = true ]; then
```
Comment (fetzerch) [inline tools/buildsteps/android-arm64-v8a/make-binary-addons]: same

File: tools/buildsteps/android-arm64-v8a/make-xbmc [hunk 5]
```diff
@@ -0,0 +1,9 @@
+WORKSPACE=${WORKSPACE:-$( cd $(dirname $0)/../../.. ; pwd -P )}
+XBMC_PLATFORM_DIR=android
+. $WORKSPACE/tools/buildsteps/defaultenv
+
+if [ "$CMAKE_BUILD" = true ]; then
```
Comment (fetzerch) [inline tools/buildsteps/android-arm64-v8a/make-xbmc]: same

File: tools/buildsteps/android-arm64-v8a/package [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+XBMC_PLATFORM_DIR=android
+. $WORKSPACE/tools/buildsteps/defaultenv
+
+if [ "x$BUILD_OBB" == "xtrue" ]; then
+  TARGET=apk-obb-unsigned
+else
+  TARGET=apk-unsigned
+fi
+
+if [ "$CMAKE_BUILD" = true ]; then
```
Comment (fetzerch) [inline tools/buildsteps/android-arm64-v8a/package]: same

File: xbmc/cores/DllLoader/exports/CMakeLists.txt [hunk 7]
```diff
@@ -21,7 +21,7 @@ elseif(NOT CORE_SYSTEM_NAME STREQUAL windows)
   add_custom_target(wrapper.def ALL nm ${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/wrapper.dir/wrapper.c.o | grep __wrap | awk '{ printf(\"%s \", \$\$3) }' | sed \"s/___wrap_/__wrap_/g\" | sed \"s/__wrap_/-Wl,-wrap,/g\" > wrapper.def)
 
   if(CORE_SYSTEM_NAME STREQUAL android)
-    add_custom_command(TARGET wrapper.def COMMAND echo \"-l$<TARGET_FILE:${APP_NAME_LC}>\" >> wrapper.def)
+    add_custom_command(TARGET wrapper.def COMMAND echo \"-L${DEPENDS_PATH}/lib/dummy-lib${APP_NAME_LC} -l${APP_NAME_LC}\" >> wrapper.def)
```
Comment (fetzerch) [inline xbmc/cores/DllLoader/exports/CMakeLists.txt]: why not `-l$<TARGET_FILE_NAME:${APP_NAME_LC}>` instead `-l${APP_NAME_LC}`

**Review by fetzerch — CHANGES_REQUESTED** [review]
I think when rebasing, you reverted the fix for cmake detecting wrapper libs (libexif) as always outdated.

File: xbmc/cores/DllLoader/exports/CMakeLists.txt (hunk 7 above)
Comment (peak3d) [inline xbmc/cores/DllLoader/exports/CMakeLists.txt]: Not working with aarch64 android linker -> full pathnames seem not to work

Comment (peak3d) [inline xbmc/cores/DllLoader/exports/CMakeLists.txt]: whats also working is not use dummy-lib, but then only with -L<cmake_binary> -lkodi and not -l<path>libkodi.so

Comment (fetzerch) [inline xbmc/cores/DllLoader/exports/CMakeLists.txt]: ah, if the extension is a problem, then leave it as is. sry

Comment (wsnipex) [inline xbmc/cores/DllLoader/exports/CMakeLists.txt]: because it is not our target_file, but dummy-libkodi.so in depends

Comment (fetzerch) [inline xbmc/cores/DllLoader/exports/CMakeLists.txt]: Sry then i completely missunderstood it. Is it worth to put a comment explaining why we need to link the dummy?

**Comment by MartijnKaijser** [comment]
@peak3d can you squash this one into the buildstep commit? 
https://github.com/MartijnKaijser/xbmc/commit/c5adda7783b5bdaf1e2c4822337da3e155b58237

**Comment by wsnipex** [comment]
here is a rebased and fixed branch https://github.com/wsnipex/xbmc/commits/aarch64
I cleaned up and squashed the buildsteps commits there

**Comment by peak3d** [comment]
Thx, @wsnipex !!!!!!!!!! gimme a few minutes

**Comment by peak3d** [comment]
git reset --hard wsnipex/aarch64, hope I did it right.....

**Comment by wsnipex** [comment]
@fetzerch ok now?

**Comment by MartijnKaijser** [comment]
Runtime tested: 
SATV with ARMv8 (64bit)
Amazon FireTV with ARMv7 (32bit)
All seems ok. Time for merge

**Pull request merged**